### ISSUE TITLE
Major bug in Transformers' masks

### ISSUE THREAD

**Issue by EmmaRenauld** [pytorch/pytorch#92554]
### 🐛 Describe the bug

In torch.nn.functional, line 5174 in current code:
![image](https://user-images.githubusercontent.com/4967417/213253672-3dfd0784-8b4c-4513-b1f0-d14a9a964612.png)

`attn_mask = attn_mask.masked_fill(key_padding_mask, float("-inf"))`

Fills `-inf` value wherever `key_padding_mask` is True. However, `key_padding_masks` actually contains -inf values instead of True values. So the padding mask is actually ignored. Replacing by:

`attn_mask = attn_mask.masked_fill(key_padding_mask < 0, float("-inf")`

Gives the correct expected mask.

### Versions

Current github code.  (master branch, Jan 18, 2023)

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @bhosmer @cpuhrsch @erichan1 @drisspg @saketh-are

**Comment by albanD** [comment]
Hi,

The doc seems to imply that `key_padding_mask` for MultiheadAttention should be a boolean or byte Tensor: https://pytorch.org/docs/stable/generated/torch.nn.quantizable.MultiheadAttention.html#torch.nn.quantizable.MultiheadAttention.forward ?
But indeed Transformer does allow `src_mask` and `tgt_mask` to be floats.

Tentatively setting high pri.

**Comment by mikekgfb** [comment]
torch.nn.MultiHeadAttention is defined to accept floats => 
https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html
"For a float mask, the mask values will be added to the attention weight."

**Comment by mikekgfb** [comment]
(at least) as early as November 2021, we issued warning about byte tensors being deprecated for torch.nn.MultiHeadAttention, e.g., here => https://github.com/pytorch/pytorch/issues/67999

```
                warnings.warn("Byte tensor for attn_mask in nn.MultiheadAttention is deprecated. Use bool tensor instead.")
                attn_mask = attn_mask.to(torch.bool)
```

**Comment by mikekgfb** [comment]
I recommend that we correct the documentation for MultiHeadAttention to reflect that byte masks were deprecated a while ago, e.g., here => https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html
as well as here => https://pytorch.org/docs/stable/generated/torch.nn.quantizable.MultiheadAttention.html#torch.nn.quantizable.MultiheadAttention.forward

**Comment by mikekgfb** [comment]
It's slightly more complicated than this?  because key_padding_mask might be either Boolean or Float.  
Are permutations allowed where one mask is Boolean and the other is FLoat, or should we require both of them to be of same type?  I would strongly advocate for the latter!  

In that case, the suggested code is sufficient (rather than requiring case splitting for kay_padding_mask being either Bool or float):
```
attn_mask = attn_mask.masked_fill(key_padding_mask < 0, float("-inf")
```

In terms of computation, I believe 
```
attn_mask = attn_mask.add(key_padding_mask) 
```
avoids allocating and instantiating a Boolean mask.

**Comment by EmmaRenauld** [comment]
I saw the deprecated warning. However, your own code, in torch.nn.modules.transformers, still creates masks using -inf.

![image](https://user-images.githubusercontent.com/4967417/213725628-d9b6ce13-6dfe-4d6a-948b-5d722abf6a0d.png)

**Comment by mikekgfb** [comment]
Presumably you had a test case that demonstrates the problem?  Can you please create a PR and submit it, so we can verify the PR fixes the issue?

**Comment by EmmaRenauld** [comment]
My code is not ready yet, I am only starting to prepare everything. I think you should not only rely on me for your tests. But yes, it will be my pleasure to prepare a PR.

**Comment by EmmaRenauld** [comment]
Hi! I saw the PR#92733. Thanks for correcting so fast.

You still have an uncorrected float usage in:

torch.nn.Transformer.generate_square_subsequent_mask.

**Comment by mikekgfb** [comment]
Can you please provide more context as to what you see as the problem?   Maybe we haven’t documented it in the doc strings clearly enough that floats are intended to be legal.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
